low-level/actuators/impl/generic/realstor_encl.py

```diff
@@ -608,7 +641,95 @@ def _get_encl_response(self, uri, request_type):
                 return response
             else:
                 logger.error(f"invalid data for {uri}")
-                return None
+                return {}
         except ValueError as err:
             logger.error(f"invalid data for {uri} {err}")
             return None
+
+    def _put_enclosure_action(self, ctrl_action, ctrl_type, resource,
+            enclosure_request):
+        severity = "informational"
+        message = "request performed successfully."
+        invalid_args = False
+        ctrl_cmd = ""
+        action_type = ""
+        if resource.startswith(("controller_a", "controller_b")):
+            ctrl_name = resource.split("controller_")[1].split(':')[0]
+        elif resource == "*":
+            ctrl_name = "both"
+        else:
+            ctrl_name = ""
+            invalid_args = True
+
+        if ctrl_action == "shutdown":
+            ctrl_cmd = f"{ctrl_action} {ctrl_name}"
+            action_type = ctrl_action
+        else:
+            ctrl_cmd = f"{ctrl_action} {ctrl_type} {ctrl_name}"
+            action_type = f"{ctrl_action} {ctrl_type}"
+
+        if (ctrl_action == "restart" and not resource.endswith(('sc', 'mc', "*"))) \
+            or (ctrl_action == "shutdown" and resource.endswith(('sc', 'mc'))) or \
+            not ctrl_cmd.endswith((' a', ' b', ' both')):

```

Comment:
in shutdown case, these controller names wont apply !?

Reply:
We have built/created the actual controller command (ctrl_cmd) in lines 664 to 669.
We are checking here if the command is not ending with a, b, or both if not then don't proceed.

Reply:
If ctrl_action == "shutdown", then following doesnt apply, still request would be marked invalid. So it should be checked with context of resource  sc, mc.

> ctrl_cmd.endswith((' a', ' b', ' both')):

Reply:
As per the discussion, I have added appropriate comments within the code.